File: backend/features/offers/offers-core/src/main/java/de/bennyboer/kicherkrabbe/offers/Notes.java
package de.bennyboer.kicherkrabbe.offers;

import jakarta.annotation.Nullable;
import lombok.AllArgsConstructor;
import lombok.Value;

import java.util.Optional;

import static de.bennyboer.kicherkrabbe.commons.Preconditions.notNull;
import static lombok.AccessLevel.PRIVATE;

@Value
@AllArgsConstructor(access = PRIVATE)
public class Notes {

    Note description;

    @Nullable
    Note contains;

    @Nullable
    Note care;

    @Nullable
    Note safety;

    public static Notes of(Note description, @Nullable Note contains, @Nullable Note care, @Nullable Note safety) {
        notNull(description, "Description note must be given");

        return new Notes(description, contains, care, safety);
    }

    public Optional<Note> getContains() {
        return Optional.ofNullable(contains);
    }

    public Optional<Note> getCare() {
        return Optional.ofNullable(care);
    }

    public Optional<Note> getSafety() {
        return Optional.ofNullable(safety);
    }

}

File: frontend/customer/src/app/offers/model/notes.ts
import { someOrNone, type Option, validateProps } from "@kicherkrabbe/shared";

export class Notes {
	readonly description: string;
	readonly contains: Option<string>;
	readonly care: Option<string>;
	readonly safety: Option<string>;

	private constructor(props: {
		description: string;
		contains: Option<string>;
		care: Option<string>;
		safety: Option<string>;
	}) {
		validateProps(props);

		this.description = props.description;
		this.contains = props.contains;
		this.care = props.care;
		this.safety = props.safety;
	}

	static of(props: {
		description: string;
		contains?: string | null;
		care?: string | null;
		safety?: string | null;
	}): Notes {
		return new Notes({
			description: props.description,
			contains: someOrNone(props.contains),
			care: someOrNone(props.care),
			safety: someOrNone(props.safety),
		});
	}
}
